core/src/main/scala/kafka/server/KafkaConfig.scala

```diff
@@ -2230,8 +2230,15 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
       validateAdvertisedListenersNonEmptyForBroker()
     } else {
       // ZK-based
-      // controller listener names must be empty when not in KRaft mode
-      require(controllerListenerNames.isEmpty, s"${KafkaConfig.ControllerListenerNamesProp} must be empty when not running in KRaft mode: ${controllerListenerNames.asJava}")
+      if (migrationEnabled) {
+        validateNonEmptyQuorumVotersForKRaft()
+        require(controllerListenerNames.nonEmpty,
```

Comment:
this seems like a bug. in migration mode you want it to be non-empty, right?

Reply:
oh, i typoed the description. the code is correct i think: `validateNonEmptyQuorumVotersForKRaft` `controllerListenerNames.nonEmpty`